File: bucket_3/1.ts
import * as grpc from '@grpc/grpc-js';
import { ChildService, FeedService, ParentService, TweetService } from './api_grpc_pb'; // Import your generated gRPC service classes
import {
    AcceptInvitationRequest, AcceptInvitationResponse,
    Child, CreateChildRequest, CreateChildResponse,
    CreateParentRequest, CreateParentResponse,
    CreateTweetRequest, CreateTweetResponse,
    DeleteTweetRequest, DeleteTweetResponse,
    Feed, FollowChildRequest, FollowChildResponse,
    InviteParentRequest, InviteParentResponse,
    LikeTweetRequest, LikeTweetResponse,
    ListChildrenRequest, ListChildrenResponse,
    ListTweetsRequest, ListTweetsResponse,
    Parent, ReplyTweetRequest, ReplyTweetResponse,
    RetweetTweetRequest, RetweetTweetResponse,
    SignUpChildRequest, SignUpChildResponse,
    Tweet, UpdateChildRequest, UpdateChildResponse,
    UpdateParentRequest, UpdateParentResponse,
    VerifyChildRequest, VerifyChildResponse,
    VerifyParentRequest, VerifyParentResponse
} from './api_pb'; // Import your generated message classes

const server = new grpc.Server();

// Implement ChildService
server.addService(ChildService, {
    createChild: (call: grpc.ServerUnaryCall<CreateChildRequest>, callback: grpc.sendUnaryData<CreateChildResponse>) => {
        // Implement the logic to create a child
        const response = new CreateChildResponse();
        // Set response properties
        callback(null, response);
    },
    verifyChild: (call: grpc.ServerUnaryCall<VerifyChildRequest>, callback: grpc.sendUnaryData<VerifyChildResponse>) => {
        // Implement the logic to verify a child
        const response = new VerifyChildResponse();
        // Set response properties
        callback(null, response);
    },
    // ... Implement other ChildService methods
});

// Implement FeedService
server.addService(FeedService, {
    listTweets: (call: grpc.ServerUnaryCall<ListTweetsRequest>, callback: grpc.sendUnaryData<ListTweetsResponse>) => {
        // Implement the logic to list tweets
        const response = new ListTweetsResponse();
        // Set response properties
        callback(null, response);
    },
    // ... Implement other FeedService methods
});

// Implement ParentService
server.addService(ParentService, {
    createParent: (call: grpc.ServerUnaryCall<CreateParentRequest>, callback: grpc.sendUnaryData<CreateParentResponse>) => {
        // Implement the logic to create a parent
        const response = new CreateParentResponse();
        // Set response properties
        callback(null, response);
    },
    verifyParent: (call: grpc.ServerUnaryCall<VerifyParentRequest>, callback: grpc.sendUnaryData<VerifyParentResponse>) => {
        // Implement the logic to verify a parent
        const response = new VerifyParentResponse();
        // Set response properties
        callback(null, response);
    },
    // ... Implement other ParentService methods
});

// Implement TweetService
server.addService(TweetService, {
    createTweet: (call: grpc.ServerUnaryCall<CreateTweetRequest>, callback: grpc.sendUnaryData<CreateTweetResponse>) => {
        // Implement the logic to create a tweet
        const response = new CreateTweetResponse();
        // Set response properties
        callback(null, response);
    },
    // ... Implement other TweetService methods
});

const serverPort = '0.0.0.0:50051';
server.bindAsync(serverPort, grpc.ServerCredentials.createInsecure(), () => {
    console.log(`Server started on ${serverPort}`);
    server.start();
});





// client side:

 CreateTweet(request: CreateTweetRequest, context: any): CreateTweetResponse {
    const newTweet = new Tweet();
    newTweet.setText(request.getText());
    newTweet.setLikes(0);
    newTweet.setOriginalTweet('');
    newTweet.setName('tweet123'); //tweet id here

    const response = new CreateTweetResponse();
    response.setTweet(newTweet);
    return response;
  }

  DeleteTweet(request: DeleteTweetRequest, context: any): DeleteTweetResponse {
    const tweetId = request.getTweetId();

    const response = new DeleteTweetResponse();
    return response;
  }

  LikeTweet(request: LikeTweetRequest, context: any): LikeTweetResponse {
    const tweetId = request.getTweetId();

    const response = new LikeTweetResponse();
    return response;
  }

  ReplyTweet(request: ReplyTweetRequest, context: any): ReplyTweetResponse {
    const parentTweetId = request.getParentTweetId();
    const replyText = request.getReplyText();

    const newReply = new Tweet();
    newReply.setText(replyText);
    newReply.setLikes(0);
    newReply.setOriginalTweet(parentTweetId);
    newReply.setName('reply123'); //reply message id

    const response = new ReplyTweetResponse();
    response.setReply(newReply);
    return response;
  }

  RetweetTweet(request: any, context: any): RetweetTweetResponse {
    const tweetId = request.getTweetId();
    //tweet logic here

    const response = new RetweetTweetResponse();
    return response;
  }

File: bucket_3/1.py
import grpc
from concurrent import futures
import api_pb2
import api_pb2_grpc

class ChildService(api_pb2_grpc.ChildServiceServicer):
    def CreateChild(self, request, context):
        # Implement the logic to create a child
        response = api_pb2.CreateChildResponse()
        return response

    def VerifyChild(self, request, context):
        # Implement the logic to verify a child
        response = api_pb2.VerifyChildResponse()
        return response

    def SignUpChild(self, request, context):
        # Implement the logic to sign up a child
        response = api_pb2.SignUpChildResponse()
        return response

    def UpdateChild(self, request, context):
        # Implement the logic to update a child
        response = api_pb2.UpdateChildResponse()
        return response

    def ListChildren(self, request, context):
        # Implement the logic to list children
        response = api_pb2.ListChildrenResponse()
        return response

    def FollowChild(self, request, context):
        # Implement the logic to follow a child
        response = api_pb2.FollowChildResponse()
        return response

class FeedService(api_pb2_grpc.FeedServiceServicer):
    def ListTweets(self, request, context):
        # Implement the logic to list tweets
        response = api_pb2.ListTweetResponse()
        return response

class ParentService(api_pb2_grpc.ParentServiceServicer):
    def CreateParent(self, request, context):
        # Implement the logic to create a parent
        response = api_pb2.CreateParentResponse()
        return response

    def VerifyParent(self, request, context):
        # Implement the logic to verify a parent
        response = api_pb2.VerifyParentResponse()
        return response

    def UpdateParent(self, request, context):
        # Implement the logic to update a parent
        response = api_pb2.UpdateParentResponse()
        return response

    def InviteParent(self, request, context):
        # Implement the logic to invite a parent
        response = api_pb2.InviteParentResponse()
        return response

    def AcceptInvitation(self, request, context):
        # Implement the logic to accept an invitation
        response = api_pb2.AcceptInvitationResponse()
        return response

class TweetService(api_pb2_grpc.TweetServiceServicer):
    def CreateTweet(self, request, context):
        # Implement the logic to create a tweet
        response = api_pb2.CreateTweetResponse()
        return response

    def DeleteTweet(self, request, context):
        # Implement the logic to delete a tweet
        response = api_pb2.DeleteTweetResponse()
        return response

    def LikeTweet(self, request, context):
        # Implement the logic to like a tweet
        response = api_pb2.LikeTweetResponse()
        return response

    def ReplyTweet(self, request, context):
        # Implement the logic to reply to a tweet
        response = api_pb2.ReplyTweetResponse()
        return response

    def RetweetTweet(self, request, context):
        # Implement the logic to retweet a tweet
        response = api_pb2.RetweetTweetResponse()
        return response

def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    api_pb2_grpc.add_ChildServiceServicer_to_server(ChildService(), server)
    api_pb2_grpc.add_FeedServiceServicer_to_server(FeedService(), server)
    api_pb2_grpc.add_ParentServiceServicer_to_server(ParentService(), server)
    api_pb2_grpc.add_TweetServiceServicer_to_server(TweetService(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    print("Server started on port 50051...")
    server.wait_for_termination()

if __name__ == '__main__':
    serve()






#Client's implementation

    def CreateTweet(self, request: CreateTweetRequest, context):
        new_tweet = api_pb2.Tweet(
            text=request.text,
            likes=0,
            original_tweet="",
            name="tweet123",  # tweet id
        )
        response = CreateTweetResponse(tweet=new_tweet)
        return response

    def DeleteTweet(self, request: DeleteTweetRequest, context):
        tweet_id = request.tweet_id
        response = DeleteTweetResponse()
        return response

    def LikeTweet(self, request: LikeTweetRequest, context):
        tweet_id = request.tweet_id
        response = LikeTweetResponse()
        return response

    def ReplyTweet(self, request: ReplyTweetRequest, context):
        parent_tweet_id = request.parent_tweet_id
        reply_text = request.reply_text
        new_reply = api_pb2.Tweet(
            text=reply_text,
            likes=0,
            original_tweet=parent_tweet_id,
            name="reply123", #reply id
        )
        response = ReplyTweetResponse(reply=new_reply)
        return response

    def RetweetTweet(self, request, context):
        tweet_id = request.tweet_id
        response = api_pb2.RetweetTweetResponse()
        return response
